hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/genesis/BenchMarkCRCStreaming.java

```diff
@@ -0,0 +1,169 @@
+ * executed on a compatible OS (ie Linux x86):
+ *
+ * ozone --jvmargs -Djava.library.path=/home/sodonnell/native genesis -b
+ *     BenchmarkCRC
+ */
+public class BenchMarkCRCStreaming {
+
+  private static int dataSize = 64 * 1024 * 1024;
+
+  @State(Scope.Thread)
+  public static class BenchmarkState {
+
+    private final ByteBuffer data = ByteBuffer.allocate(dataSize);
+
+    @Param({"512", "1024", "2048", "4096", "32768", "1048576"})
+    private int checksumSize;
+
+    @Param({"pureCRC32", "pureCRC32C", "hadoopCRC32C", "hadoopCRC32",
+        "zipCRC32", "zipCRC32C", "nativeCRC32", "nativeCRC32C"})
+    private String crcImpl;
+
+    private ChecksumByteBuffer checksum;
+
+    public ChecksumByteBuffer checksum() {
+      return checksum;
+    }
+
+    public String crcImpl() {
+      return crcImpl;
+    }
+
+    public int checksumSize() {
+      return checksumSize;
+    }
+
+    @Setup(Level.Trial)
+    public void setUp() {
+      switch (crcImpl) {
+      case "pureCRC32":
+        checksum = new PureJavaCrc32ByteBuffer();
+        break;
+      case "pureCRC32C":
+        checksum = new PureJavaCrc32CByteBuffer();
+        break;
+      case "hadoopCRC32":
+        checksum = new ChecksumByteBufferImpl(new PureJavaCrc32());
+        break;
+      case "hadoopCRC32C":
+        checksum = new ChecksumByteBufferImpl(new PureJavaCrc32C());
+        break;
+      case "zipCRC32":
+        checksum = new ChecksumByteBufferImpl(new CRC32());
+        break;
+      case "zipCRC32C":
+        try {
+          checksum = new ChecksumByteBufferImpl(
+              ChecksumByteBufferImpl.Java9Crc32CFactory.createChecksum());
+        } catch (Throwable e) {
+          throw new RuntimeException("zipCRC32C is not available pre Java 9");
+        }
+        break;
+      case "nativeCRC32":
+        if (NativeCRC32Wrapper.isAvailable()) {
+          checksum = new ChecksumByteBufferImpl(new NativeCheckSumCRC32(
+              NativeCRC32Wrapper.CHECKSUM_CRC32, checksumSize));
+        } else {
+          throw new RuntimeException("Native library is not available");
+        }
+        break;
+      case "nativeCRC32C":
+        if (NativeCRC32Wrapper.isAvailable()) {
+          checksum = new ChecksumByteBufferImpl(new NativeCheckSumCRC32(
+              NativeCRC32Wrapper.CHECKSUM_CRC32C, checksumSize));
+        } else {
+          throw new RuntimeException("Native library is not available");
+        }
+        break;
+      default:
+      }
+      data.clear();
+      data.put(RandomUtils.nextBytes(data.remaining()));
+    }
+  }
+
+  @Benchmark
+  @Threads(1)
+  @Warmup(iterations = 3, time = 1000, timeUnit = MILLISECONDS)
+  @Fork(value = 1, warmups = 0)
+  @Measurement(iterations = 5, time = 2000, timeUnit = MILLISECONDS)
+  @BenchmarkMode(Mode.Throughput)
+  public void runCRC(Blackhole blackhole, BenchmarkState state) {
+    ByteBuffer data = state.data;
+    data.clear();
```

Comment:
Why calling clearing the data?  Typo?

Reply:
clear does not actually alter the buffer contents, it only sets the position to zero and the limit to the capacity, getting the buffer read for a new read / write. I guess I don't need the clear here as I set position and limit on each pass around the loop, so I think I can remove line safely.

Reply:
You are right -- clear() does not really clear the buffer.  Thanks.